packages/docusaurus/src/client/RouteAnnouncerWrapper.tsx

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React from 'react';
+import {Portal} from './exports/Portal';
+import type {RouteComponentProps} from 'react-router-dom';
+
+/* eslint-disable @typescript-eslint/ban-ts-comment */
+
+class RouteAnnouncerWrapper extends React.Component {
+  constructor({
+    children,
+    location,
+  }: {
+    children: React.ReactNode;
+  } & RouteComponentProps) {
+    super({children, location});
+    this.state = {
+      routeAnnouncement: '',
+    };
+  }
+
+  override componentDidUpdate(): void {
```

Comment:
Just to confirm: the reason you are using `componentDidUpdate` (and hence a class component) is on purpose, because you want the announcement to happen synchronously, correct?

Reply:
I am using this because the route announcer can work correctly, even on slow networks. I tested on my side, and it works OK (like Gatsby 4 did)

Reply:
I'm also using this lifecycle because it doesn't trigger on the initial pre-render

Reply:
`useEffect` would work just the same, from some quick eyeballing? Have you tried?

Reply:
> doesn't trigger on the initial pre-render

pre-rendering, or hydration? Even if you want to skip hydration, you can use a `mounted` ref to bypass this. We only fall back to class component if function component can't work at all (e.g. `componentDidCatch`).

Reply:
Actually, scratch it. I'm fine with what we have now; `componentDidUpdate` seems like a good fit.